torch/csrc/jit/tensorexpr/mem_dependency_checker.cpp

```diff
@@ -0,0 +1,1261 @@
+    currentScope_->shadowedVarBounds[var] = knownVarBounds_[var];
+  }
+
+  currentScope_->localVars.insert(var);
+  knownVarBounds_[var] = {v->value(), v->value()};
+}
+
+// Don't support AtomicAdd yet, it's a bit more complex since it's both a read
+// and a write. It's only inserted during Cuda codegen so this should be okay.
+void MemDependencyChecker::visit(const AtomicAdd* v) {
+  throw std::runtime_error("MemDependencyChecker AtomicAdd unimplemented");
+}
+
+void MemDependencyChecker::updateWriteHistory(
+    std::list<BoundRelationship>& writeHistory,
+    const std::shared_ptr<AccessInfo>& info,
+    int latestAccessToClose,
+    bool insert) {
+  bool isWrite = info->isWrite();
+
+  for (auto it = writeHistory.begin(); it != writeHistory.end();) {
+    auto& indexBounds = it->first;
+    std::shared_ptr<AccessInfo> other = it->second;
+    if (info->hasDependency(other)) {
+      ++it;
+      continue;
+    }
+
+    OverlapKind overlap = overlaps(indexBounds, info->bounds());
+
+    if (overlap == NoOverlap) {
+      ++it;
+      continue;
+    }
+
+    // Only writes can close open accesses.
+    if (!isWrite) {
+      info->addDependency(other);
+      other->addDependent(info);
+      ++it;
+      continue;
+    }
+
+    // If we're not closing accesses we can stop here.
+    if ((int)other->id() > latestAccessToClose) {
+      ++it;
+      continue;
+    }
+
+    if (overlap == TotalOverlapB) {
+      // Total overlap is easy - the new access totally replaces the old.
+      it = writeHistory.erase(it);
+    } else {
+      // The new write partially overlaps a previous write. We want to keep
+      // both, but only track the unconvered part of the earlier write.
+
+      // Determine the slices of the earlier bound not covered by info.
+      auto newBounds =
+          subtractIndicesBounds(indexBounds, info->bounds(), overlap);
+
+      // Erase the old slice.
+      it = writeHistory.erase(it);
+
+      // Add all new slices.
+      for (auto& b : newBounds) {
+        it = writeHistory.insert(it, std::make_pair(b, other));
+      }
+      it++;
+    }
+  }
+
+  if (insert && isWrite) {
+    writeHistory.emplace_back(std::make_pair(info->bounds(), info));
+  }
+}
+
+void MemDependencyChecker::mergeScope(
+    const std::shared_ptr<Scope>& child,
+    const std::shared_ptr<Scope>& parent,
+    bool closeOverlapped) {
+  if (child->accesses_.empty()) {
+    return;
+  }
+
+  int latestAccessToClose =
+      closeOverlapped ? (int)child->accesses_.front()->id() : -1;
+
+  // Update dependencies, but don't add new open writes yet.
+  for (auto& info : child->accesses_) {
+    auto& writeHistory = parent->openWrites_[info->var()];
+    updateWriteHistory(writeHistory, info, latestAccessToClose, false);
+  }
+
+  // Copy open writes up.
+  for (auto& pair : child->openWrites_) {
+    const Var* var = pair.first;
+    for (auto& rel : pair.second) {
+      auto& writeHistory = parent->openWrites_[var];
```

Comment:
Here it looks like you do want the side effect of opening writes in the parent. It's so confusing though - I'd rather not use indexing operator anywhere than think whether the side effect is correct or a bug.

Reply:
It's not always easy when you do want the side effect and also a reference to the value. Would a comment in these cases satisfy you?

Reply:
You would get an iterator to the key and value pair. IMO that's just as easy, only more verbose.

I don't know if a comment here would be as good as changing the code. Yes, I would be satisfied since I already made a pass through this code, but I'm not so sure about other readers.

Reply:
It's a lot more verbose. You're replacing:
```
auto& writeHistory = parent_->openWrites_[var];
updateWriteHistory(writeHistory, ...);
```
with
```
auto it = parent_->openWrites_.find(var);
if (it == parent_->openWrites_.end()) {
  auto res = parent_->openWrites_.emplace(var, {});
  TORCH_INSERT_INTERNAL(res.second);
  it = res.first;
}
updateWriteHistory(it->second, ...)
```

This reads much worse to me personally, but not a hill I'm willing to die on :)

Reply:
If we consistently use the side-effecting behavior, comments suffice I think. It's not that surprising to use the indexing operator if we do want the side effects.